mantidproject/mantid
# Robustness improvements to Python fit functions

**Description of work**

This work fixes issues where users could supply incorrect Python function class definitions and mantid would crash rather than checking and throwing an appropriate error.

**To test:**

* Try out the instructions in the issue
* Try out variations of `function1D`, i.e. wrong return type, misspelled to try and get it to crash. 
* Try the example solutions from the [advanced training course](https://www.mantidproject.org/Extending_Mantid_With_Python:_Exercise_6_Solution) with and without the derivatives. The peak function types without a derivative should now work also. 

Fixes #22331, fixes #16841. fixes #7914

*Release notes have been added.* 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
